README.md
# Lab_2
Лабораторная работа No 2. Реализация NetFlow-сенсора

## Описание

Сенсор scanflowd предназначен для сбора и отправки сетевой статистики по протоколу NetFlow v.9

Сенсор формирует поток по следующим общим критериям пакетов:

1. протокол
2. тип обслуживания (IPv4)
3. адрес источника
4. адрес назначения
5. порт источника или тип ICMP
6. порт назначения или код ICMP

Порт сенсора: 9994

В статистике потоков содержется следующая информация:

1. суммарное число байт
2. суммарное число пакетов
3. время первого пакета
4. время последнего пакета

Если поток протокола TCP, то все флаги всех пакетов суммируются

Данные последнего пакета:

5. MAC-адрес источника
6. MAC-адрес назначения
7. ID IPv4

## Сборка

Для сборки программ вам потребуется пакет GNU GCC.
В принципе, все что нужно, это:
1. Скачать средства сборки: `sudo apr-get install gcc`
2. Перейти в директорию с Makefile и запустить make `make install`

Чтобы в текущей директории были созданы исполняемые файлы, исплользуйте `make all`

## Запуск

Для запуска программы необходимы права суперпользователя. Синопсис:

`sudo scanerflowd [interface name (enp*s* ...)] [ip коллектора]:[порт]`

Чтобы узнать доступные сетевые интерфейсы, можно воспользоваться командой `networkctl`

Для тестирования работоспособности программы можно использовать утилиту nfcapd как коллектор NetFlow

Запуск коллектора:

`sudo nfcapd -l ~/stat/ -b 127.0.0.1`

Запуск сенсора:

`sudo ./scanflowd [interface name (enp*s* ...)] 127.0.0.1:9995`

## Авторство и лицензия

Разработчик: Филатов И.А.
e-mail: f2000_99@mail.ru

Лицензия GNU GPL v.3


debian/changelog
lab-work-2 (1.0.1-1) unstable; urgency=medium

  * Initial release

 -- Filatov Ivan <f2000_99@mail.ru>  Sun, 31 Jul 2022 17:03:52 +0300


src/scanflowd/net_flow.h
#ifndef NET_FLOW_H
#define NET_FLOW_H

#include <netinet/ether.h>
#include <netinet/ip.h>
#include <netinet/udp.h>
#include <netinet/tcp.h>
#include <netinet/icmp6.h>
#include <time.h>

// заголовки ip и фрейма
#define FRAME_HSIZE 14
#define IP_HSIZE (FRAME_HSIZE + sizeof(struct iphdr))

// информация о сенсоре
struct exporter {
    unsigned int flows;         // число потоков

    // timeout
    unsigned short f_active;    // отправка данных каждые X sec
    unsigned short f_inactive;  // интервал неактивности
    unsigned short input_snmp;  // индекс сетевого интерфейса
    char *if_name;              // имя сетевого интерфейса

    // для заголовка
    unsigned int sys_up_time;   // время работы сенсера
    unsigned int source_id;     // индекс сетевого интерфейса
    unsigned int sq_number;     // число отправленных netflow пакетов
};

// поток netflow
struct flow {
    // определение потока
    char protocol;
    char tos;
    unsigned int ip_s;
    unsigned int ip_d;   
    unsigned short port_s_type; // порт отправителя или тип
    unsigned short port_d_code; // порт получателя или код

    // переменные потока
    unsigned int in_bytes;
    unsigned int in_pkts;

    unsigned int last_swtch;
    unsigned int first_swtch;

    char tcp_flags;

    // данные с последнего пакета
    unsigned int ipv4_ident;
    char in_src_mac[6];
    char in_dst_mac[6];
};

struct flow_list {
#define FL_LIST_SIZE 64
    struct flow data[FL_LIST_SIZE];
    unsigned short free_id;     // свободный id
};


// управление списком потоков

// добавление потока
void set_flow(struct flow *a, struct flow *b);

// обновление списка потоков
void flow_list_update(struct flow_list *fl_list, unsigned short f_inactive);

// обновление данных потока
void flow_update(struct flow *net_fl, const char *buff);

// поиск потока
unsigned short find_flow_id(struct flow_list *fl_list, struct flow *net_fl);


// создание пакетов

unsigned short new_nf_tpl(char *nf_packet, struct exporter *sensor);

unsigned int new_nf_data(char *nf_packet, struct flow *fl,
                         struct exporter *sensor);

#endif // NET_FLOW_H


src/scanflowd/scanflowd.c
#include <stdio.h>
#include <stdlib.h>
#include <signal.h>
#include <string.h>
#include <netinet/in.h>
#include <unistd.h>

#include "safe_functions.h"
#include "net_flow.h"

// обычно, максимальный размер ip пакета до 576
#define BUFF_SIZE 576

// общие данные
struct argum {
    int listening_sock;
    int collect_sock;
    char buff[BUFF_SIZE];
    struct sockaddr_in collector_sa;
    struct exporter sensor;
    struct flow_list fl_list;
};

void *listening_thread(void *argum)
{
    struct argum *a = argum;
    struct flow net_fl;
    unsigned short flow_id;
    size_t buff_len = sizeof(a->buff);

    while (1) {
        net_fl.in_bytes = recv(a->listening_sock, a->buff, buff_len, 0);

        struct iphdr *ip = (struct iphdr *) a->buff + FRAME_HSIZE;

        // проверка пакета и заполнение
        switch (ip->protocol) {
        case 1:
        case 6:
        case 17:
            net_fl.protocol = ip->protocol;
            net_fl.tos = ip->tos;
            net_fl.ipv4_ident = ip->id;
            net_fl.ip_s = ip->saddr;
            net_fl.ip_d = ip->daddr;
            if (ip->protocol == 1) {
                struct icmp6_hdr *icmp = (struct icmp6_hdr *) a->buff
                        + IP_HSIZE;
                net_fl.port_s_type = icmp->icmp6_type;
                net_fl.port_d_code = icmp->icmp6_code;
            }
            else {
                struct udphdr *udp = (struct udphdr *) a->buff + IP_HSIZE;
                net_fl.port_s_type = udp->source;
                net_fl.port_d_code = udp->dest;
            }

            // поиск потока
            flow_id = find_flow_id(&a->fl_list, &net_fl);
            if (flow_id == 65500){
                // потока нет
                flow_id = a->fl_list.free_id;
                set_flow(&a->fl_list.data[flow_id], &net_fl);
                // защита от переполнения
                if (a->fl_list.free_id < FL_LIST_SIZE - 1) {
                    a->sensor.flows++;
                    a->fl_list.free_id++;
                }
            }
            // обновление потока
            flow_update(&a->fl_list.data[flow_id], a->buff);
            break;
        default:
            break;
        }
    }
}

void *secondary_tread(void *argum)
{
    struct argum *a = argum;
    struct sockaddr_in sa = a->collector_sa;
    char nf_packet[1500];   // максимум для пакета 1500 байт
    unsigned int bytes;
    unsigned int time_of_send = time(0);
    unsigned short i;

    // отправка шаблонов
    a->sensor.sq_number++;
    bytes = new_nf_tpl(nf_packet, &a->sensor);
    bytes = sendto(a->collect_sock, nf_packet, bytes, 0,
           (struct sockaddr *) &sa, sizeof(sa));
    memset(nf_packet, 0, bytes);

    // отправка данных
    while (1) {
        // удаление неактивных потоков
        flow_list_update(&a->fl_list, a->sensor.f_inactive);

        // пришло время отправки
        if ((unsigned short) difftime(time(0), time_of_send) ==
                a->sensor.f_active) {
            for (i = 0; i < FL_LIST_SIZE; i++) {
                if (a->fl_list.data[i].protocol > 0) {
                    a->sensor.sq_number++;
                    bytes = new_nf_data(nf_packet, &a->fl_list.data[i],
                                        &a->sensor);

                    sendto(a->collect_sock, nf_packet, bytes, 0,
                           (struct sockaddr *) &sa, sizeof(sa));
                    memset(nf_packet, 0, bytes);
                    time_of_send = time(0);
                }
            }
        }
    }
}

static int working = 1;
void safe_exit(int sig){
    working = 0;
}

int main (int argc, char **argv)
{
    if (argc < 2) {
        puts("scanerflowd [interface name (enp*s* ...)] [collector ip]:[port]");
        exit(EXIT_FAILURE);
    } 

    int listening_socket = create_raw_socket();
    int collect_socket = create_udp_socket();

    // определение индекса сетевого интерфейса
    struct ifreq if_req;
    strcpy(if_req.ifr_ifrn.ifrn_name, argv[1]);
    set_data_for_ifreq(listening_socket, &if_req);

    struct sockaddr_ll sa_ll = {
        .sll_family = AF_PACKET,
        .sll_pkttype = PACKET_OTHERHOST,
        .sll_protocol = htons(ETH_P_ALL),
        .sll_ifindex = if_req.ifr_ifindex
    };
    bind_socket(listening_socket, (struct sockaddr *) &sa_ll, sizeof(sa_ll));

    uint32_t ip;
    inet_pton(AF_INET, if_req.ifr_ifru.ifru_addr.sa_data, &ip);
    struct sockaddr_in sa_cl = {
        .sin_family = AF_INET,
        .sin_port = 9994,           // порт сенсора по умолчанию
        .sin_addr.s_addr = ip       // ip адрес сетевого интерфейса
    };
    bind_socket(collect_socket, (struct sockaddr *) &sa_cl, sizeof(sa_cl));

    // обработка адреса коллектора
    char *port_str = strchr(argv[2], ':');
    *port_str = 0;
    port_str++;
    inet_pton(AF_INET, argv[2], &ip);
    sa_cl.sin_port = htons(atoi(port_str)); // порт коллектора

    // подготовка потоков
    pthread_t list_thread_id, secd_thread_id;
    pthread_attr_t thr_attr;
    pthread_attr_init(&thr_attr);

    // аргументы для передачи в поток
    struct argum a = {
        .buff = {0},
        .listening_sock = listening_socket,
        .collect_sock = collect_socket,
        .collector_sa = sa_cl,
        .sensor = {
            .if_name = argv[1],
            .input_snmp = if_req.ifr_ifindex,
            .source_id = if_req.ifr_ifindex,
            .f_active = 10, // 30
            .f_inactive = 15,
            .flows = 0,
            .sq_number = 0,
            .sys_up_time = time(0)
        },
        .fl_list = {
            .data = {0},
            .free_id = 0,
        }
    };

    // потоки
    pthread_create(&list_thread_id, &thr_attr, listening_thread, &a);
    pthread_create(&secd_thread_id, &thr_attr, secondary_tread, &a);

    // ctrl+c
    signal(SIGINT, safe_exit);

    while (working) {}

    pthread_cancel(list_thread_id);
    pthread_cancel(secd_thread_id);

    close(collect_socket);
    close(listening_socket);

    exit(EXIT_SUCCESS);
}



src/scanflowd/net_flow.c
#include "net_flow.h"

/* из за разного порядка байт по разному работает memcpy
 * функции перевода int/short -> массив char
 * size_t n - число переводимых элементов
*/

void memcpy_sh(char *dist, unsigned short *src, size_t n)
{
    size_t i, j = 0;
    for (i = 0; j < n; i+=2, j++) {
        dist[i + 1] = src[j] & 0xFF;
        dist[i] = (src[j] & 0xFF00) >> 8;
    }
}

void memcpy_int(char *dist, unsigned int *src, size_t n) // Решить проблему!!
{
    unsigned short a;
    size_t i, j = 0;
    for (i = 0; j < n; i+=4, j++) {
        a = (src[j] & 0xFFFF);
        memcpy_sh(dist + i + 2 , &a, 1);
        a = (src[j] & 0xFFFF0000) >> 16;
        memcpy_sh(dist + i, &a, 1);
    }
}

// управление списком потоков

void set_flow(struct flow *a, struct flow *b)
{
    a->protocol = b->protocol;
    a->tos = b->tos;
    a->ipv4_ident = b->ipv4_ident;
    a->ip_s = b->ip_s;
    a->ip_d = b->ip_d;
    a->port_s_type = b->port_s_type;
    a->port_d_code = b->port_d_code;
    a->tcp_flags = b->tcp_flags;
}

void flow_list_update(struct flow_list *fl_list, unsigned short f_inactive)
{
    unsigned short i;
    for (i = 0; i < FL_LIST_SIZE; i++) {
        if ((time(0) - fl_list->data[i].last_swtch) == f_inactive) {
            fl_list->data[i].protocol = 0; // поток недействителен
        }
    }
}

void flow_update(struct flow *net_fl, const char *buff)
{
    struct ethhdr *eth = (struct ethhdr *) buff;
    struct iphdr *ip = (struct iphdr *) buff + FRAME_HSIZE;

    net_fl->in_pkts++;
    net_fl->in_bytes += ip->tot_len;

    memcpy(net_fl->in_src_mac, eth->h_source, 6);
    memcpy(net_fl->in_dst_mac, eth->h_dest, 6);

    if (net_fl->first_swtch == 0) {
        net_fl->first_swtch = time(0);
    }
    net_fl->last_swtch = time(0);

    if (ip->protocol == 6) {
        struct tcphdr *tcp = (struct tcphdr *) buff + IP_HSIZE;
        net_fl->tcp_flags |= tcp->th_flags;
    }
}

unsigned short find_flow_id(struct flow_list *fl_list, struct flow *net_fl)
{
    // поиск потока в списке
    unsigned short i;
    for (i = 0; i < FL_LIST_SIZE; i++) {
        if (
                net_fl->protocol == fl_list->data[i].protocol &&
                net_fl->tos == fl_list->data[i].tos &&
                net_fl->ip_s == fl_list->data[i].ip_s &&
                net_fl->ip_d == fl_list->data[i].ip_d &&
                net_fl->port_s_type == fl_list->data[i].port_s_type &&
                net_fl->port_d_code == fl_list->data[i].port_d_code
                ) {
            return i;
        }
    }
    // невозможный индекс, означающий, то потока нет
    return 65500;
}


// создание элементов пакета

unsigned short create_nf_tpl(unsigned short *nf_tpl, unsigned short template_ID,
                   unsigned short ifname_size, char proto)
{
    // для шаблона
    unsigned short field_count = 0;
    unsigned short data[6];

    if (proto == 1) {
        // ICMP
        data[0] = 32;   data[1] = 2;    // ICMP_TYPE icmp [type| code] 32bit
        field_count++;
    }
    else if (proto == 6) {
        // TCP
       data[0] = 7;     data[1] = 2;    // L4_SRC_PORT
       data[2] = 11;    data[3] = 2;    // L4_DST_PORT
       data[4] = 6;     data[5] = 1;    // TCP_FLAGS
       field_count += 3;
    }
    else {
        // UDP
        data[0] = 7;     data[1] = 2;    // L4_SRC_PORT
        data[2] = 11;    data[3] = 2;    // L4_DST_PORT
        field_count += 2;
    }

    unsigned short data_size = field_count * 2;
    field_count += 16;
    unsigned short lenght = (field_count + 2) * 4;

    nf_tpl[0] = 0; nf_tpl[1] = lenght,
    nf_tpl[2] = template_ID; nf_tpl[3] = field_count,
        // все что 4 байта
    nf_tpl[4] = 1; nf_tpl[5] = 4;    // IN_BYTES
    nf_tpl[6] = 2; nf_tpl[7] = 4;    // IN_PKTS
    nf_tpl[8] = 3; nf_tpl[9] = 4;    // FLOWS
    nf_tpl[10] = 21; nf_tpl[11] = 4;    // LAST_SWITCHED
    nf_tpl[12] = 22; nf_tpl[13] = 4;    // FIRST_SWITCHED
    nf_tpl[14] = 8; nf_tpl[15] = 4;    // IPV4_SRC_ADDR
    nf_tpl[16] = 12; nf_tpl[17] = 4;    // IPV4_DST_ADDR
        // все что 2 байта и т.д.
    nf_tpl[18] = 36; nf_tpl[19] = 2;    // FLOW_ACTIVE_TIMEOUT
    nf_tpl[20] = 37; nf_tpl[21] = 2;    // FLOW_INACTIVE_TIMEOUT
    nf_tpl[22] = 10; nf_tpl[23] = 2;    // INPUT_SNMP
    nf_tpl[24] = 54; nf_tpl[25] = 2;    // IPV4_IDENT

    nf_tpl[26] = 5; nf_tpl[27] = 1;    // SRC_TOS
    nf_tpl[28] = 4; nf_tpl[29] = 1;    // PROTOCOL
        // MAC
    nf_tpl[30] = 56; nf_tpl[31] = 6;    // IN_SRC_MAC
    nf_tpl[32] = 57; nf_tpl[33] = 6;    // IN_DST_MAC

    nf_tpl[34] = 82; nf_tpl[35] = ifname_size;  // IF_NAME

    // дописание в зависимости от протокола
    for (unsigned short i = 0; i < data_size + 1; ++i) {
        nf_tpl[36 + i] = data[i];
    }

    return lenght;
}

unsigned short create_nf_data(char *nf_data, unsigned short template_ID,
                    struct flow *net_fl, struct exporter *sensor)
{
    char data[5], data_len;

    if (net_fl->protocol != 1) {
        memcpy(data, &net_fl->port_s_type, 2);
        memcpy(data + 2, &net_fl->port_d_code, 2);
        data_len = 4;
        if (net_fl->protocol == 6) {    // TCP
            data[4] = net_fl->tcp_flags;
            data_len++;
        }
    }
    else {
        memcpy(data, &net_fl->port_s_type, 1);
        memcpy(data + 1, &net_fl->port_d_code, 1);
        data_len = 2;
    }

    unsigned int d_32[7] = {
        // все что 4 байта
        net_fl->in_bytes,
        net_fl->in_pkts,
        sensor->flows,
        net_fl->last_swtch,
        net_fl->first_swtch,
        net_fl->ip_s,
        net_fl->ip_d
    };

    unsigned short d_16[4] = {
        // все что 2 байта и т.д.
        sensor->f_active,
        sensor->f_inactive,
        sensor->input_snmp,
        net_fl->ipv4_ident
    };

    unsigned char d_8[2] = {
        net_fl->tos,
        net_fl->protocol
    };

    // IF_NAME
    unsigned short if_name_size = sizeof(sensor->if_name);

    unsigned short lenght = 4 + 28 + 8 + 2 + 12 + if_name_size + data_len;

    memcpy_sh(nf_data, &template_ID, 1);
    memcpy_sh(nf_data + 2, &lenght, 1);
    memcpy_int(nf_data + 4, d_32, 7);
    memcpy_sh(nf_data + 32, d_16, 4);
    memcpy(nf_data + 40, d_8, 2);
    memcpy(nf_data + 42, net_fl->in_src_mac, 6);
    memcpy(nf_data + 48, net_fl->in_dst_mac, 6);

    memcpy(nf_data + 54, sensor->if_name, if_name_size);
    memcpy(nf_data + 54 + if_name_size, data, data_len);

    return lenght;
}

void create_nf_header(unsigned int *nf_header, unsigned short count,
                      struct exporter *sensor, unsigned int source_id)
{
    nf_header[0] = (9 << 16) + count;   // версия + количество flow_sets
    nf_header[1] = sensor->sys_up_time;
    nf_header[2] = time(0);             // время отправки пакета
    nf_header[3] = sensor->sq_number;   // число пакетов
    nf_header[4] = source_id;
}

// создание пакетов NetFlow

unsigned short new_nf_tpl(char *nf_packet, struct exporter *sensor)
{
    unsigned short nf_tpl[42], bytes = 20, tpl_bytes = 0;

    // заголовок
    unsigned int nf_header[20];
    create_nf_header(nf_header, 3, sensor, sensor->source_id);
    memcpy_int(nf_packet, nf_header, 5);

    // ICMP
    tpl_bytes = create_nf_tpl(nf_tpl, 256, sizeof (sensor->if_name), 1);
    memcpy_sh(nf_packet + bytes, nf_tpl, tpl_bytes);
    bytes += tpl_bytes;

    // UDP
    tpl_bytes = 0;
    tpl_bytes = create_nf_tpl(nf_tpl, 257, sizeof (sensor->if_name), 17);
    memcpy_sh(nf_packet + bytes, nf_tpl, tpl_bytes);
    bytes += tpl_bytes;

    // TCP
    tpl_bytes = 0;
    tpl_bytes = create_nf_tpl(nf_tpl, 258, sizeof (sensor->if_name), 6);
    memcpy_sh(nf_packet + bytes, nf_tpl, tpl_bytes);

    return bytes + tpl_bytes;
}


unsigned int new_nf_data(char *nf_packet, struct flow *fl,
                         struct exporter *sensor)
{
    unsigned short bytes = 20, flow_sets = 0;
    unsigned int nf_header[20];
    create_nf_header(nf_header, flow_sets, sensor, sensor->source_id);
    memcpy_int(nf_packet, nf_header, 5);

    // запись по шаблонам
    if (fl->protocol == 1) {
        bytes += create_nf_data(nf_packet + 20, 256, fl, sensor);
    }
    else if (fl->protocol == 17) {
        bytes += create_nf_data(nf_packet + 20, 257, fl, sensor);
    }
    else if (fl->protocol == 6){
        bytes += create_nf_data(nf_packet + 20, 258, fl, sensor);
    }

    return bytes;
}
